docs/dev/how-to/estimate-gas.md

```diff
@@ -0,0 +1,399 @@
+
+const wallet = new Wallet(TEST_PRIVATE_KEY, l2Provider, l1Provider);
+
+async function sendMessageToL1(text: string) {
+  console.log(`Sending message to L1 with text ${text}`);
+  const textBytes = ethers.utils.toUtf8Bytes(MESSAGE);
+
+  const messengerContract = new ethers.Contract(utils.L1_MESSENGER_ADDRESS, utils.L1_MESSENGER, wallet);
+  const tx = await messengerContract.sendToL1(textBytes);
+  await tx.wait();
+  console.log("L2 trx hash is ", tx.hash);
+  return tx;
+}
+
+async function getL2MessageProof(blockNumber: ethers.BigNumberish) {
+  console.log(`Getting L2 message proof for block ${blockNumber}`);
+  return await l2Provider.getMessageProof(blockNumber, wallet.address, ethers.utils.keccak256(ethers.utils.toUtf8Bytes(MESSAGE)));
+}
+
+async function proveL2MessageInclusion(l1BatchNumber: ethers.BigNumberish, proof: any, trxIndex: number) {
+  const zkAddress = await l2Provider.getMainContractAddress();
+
+  const mailboxL1Contract = new ethers.Contract(zkAddress, utils.ZKSYNC_MAIN_ABI, l1Provider);
+  // all the information of the message sent from L2
+  const messageInfo = {
+    txNumberInBlock: trxIndex,
+    sender: wallet.address,
+    data: ethers.utils.toUtf8Bytes(MESSAGE),
+  };
+
+  console.log(`Retrieving proof for batch ${l1BatchNumber}, transaction index ${trxIndex} and proof id ${proof.id}`);
+
+  const res = await mailboxL1Contract.proveL2MessageInclusion(l1BatchNumber, proof.id, messageInfo, proof.proof);
+
+  return res;
+}
+
+/**
+ * Full end-to-end of an L2-L1 messaging with proof validation.
+ * Recommended to run in 3 steps:
+ * 1. Send message.
+ * 2. Wait for transaction to finalize and block verified
+ * 3. Wait for block to be verified and validate proof
+ */
+async function main() {
+  // Step 1: send message
+  const l2Trx = await sendMessageToL1(MESSAGE);
+
+  console.log("Waiting for transaction to finalize...");
+
+  // Step 2: waiting to finalize can take a few minutes.
+  const l2Receipt = await l2Trx.waitFinalize();
+
+  // Step 3: get and validate proof (block must be verified)
+  const proof = await getL2MessageProof(l2Receipt.blockNumber);
+
+  console.log(`Proof is: `, proof);
+
+  const { l1BatchNumber, l1BatchTxIndex } = await l2Provider.getTransactionReceipt(l2Receipt.transactionHash);
+
+  console.log("L1 Index for Tx in block :>> ", l1BatchTxIndex);
+
+  console.log("L1 Batch for block :>> ", l1BatchNumber);
+
+  // IMPORTANT: This method requires that the block is verified
+  // and sent to L1!
+  const result = await proveL2MessageInclusion(
+    l1BatchNumber,
+    proof,
+    // @ts-ignore
+    l1BatchTxIndex
+  );
+
+  console.log("Result is :>> ", result);
+  process.exit();
+}
+
+try {
+  main();
+} catch (error) {
+  console.error(error);
+}
+```
+@tab Python
+```sh
+// In progress. Check back later.
+```
+@tab Go
+```sh
+// In progress. Check back later.
+```
+:::
+
+### Result
+
+5. The return value is the `keccak256` hash of the message bytes.
```

Comment:
`proveL2MessageInclusion` returns `true` or `false` depending on whether the proof is correct. (if that's what you mean by `result` here

Reply:
Thanks :) I think I probably need to rework the examples a little bit.

Reply:
I have amended the returns in this section.